metadata_writer_nne.c

```diff
@@ -94,132 +309,725 @@ static uint8_t md_nne_handle_gps_event(struct md_writer_nne *mwn,
+        case NNE_TYPE_INT16:
+            return md_nne_add_json_key_value_int(obj, key, value.u.v_int16);
+        case NNE_TYPE_UINT16:
+            return md_nne_add_json_key_value_int(obj, key, value.u.v_uint16);
+        case NNE_TYPE_INT32:
+            return md_nne_add_json_key_value_int(obj, key, value.u.v_int32);
+        case NNE_TYPE_UINT32:
+            return md_nne_add_json_key_value_int(obj, key, value.u.v_uint32);
+        case NNE_TYPE_STRING:
+            return md_nne_add_json_key_value_string(obj, key, value.u.v_str);
+    }
+    return RETVAL_FAILURE;
+}
+
+static void md_nne_send_message(struct md_writer_nne *mwn,
+        struct nne_message* msg)
+{
+    static const char *NNE_MESSAGE_TYPE_STR[] = {
+        "event",
+        "bins-1min"
+    };
+
+    static const char *NNE_MESSAGE_SOURCE_STR[] = {
+        "report",
+        "query"
+    };
+
+    struct json_object* obj = json_object_new_object();
+
+    md_nne_add_json_key_value_string(obj, "type", NNE_MESSAGE_TYPE_STR[msg->type]);
+    md_nne_add_json_key_value_int(obj, "ts", msg->tstamp);
+    md_nne_add_json_key_value_string(obj, "node", msg->node);
+
+    if (msg->mccmnc == 0)
+        md_nne_add_json_key_value_null(obj, "mccmnc");
+    else
+        md_nne_add_json_key_value_int(obj, "mccmnc", msg->mccmnc);
+
+    md_nne_add_json_key_value_string(obj, "key", msg->key);
+    md_nne_add_json_key_value(obj, "value", msg->value);
+
+    if (msg->extra == NULL)
+        md_nne_add_json_key_value_null(obj, "extra");
+    else
+        md_nne_add_json_key_value_string(obj, "extra", msg->extra);
+
+    md_nne_add_json_key_value_string(obj, "source", NNE_MESSAGE_SOURCE_STR[msg->source]);
+
+    if (msg->type == NNE_MESSAGE_TYPE_BINS1MIN)
+        md_nne_add_json_key_value_int(obj, "delta", msg->delta);
+
+    META_PRINT_SYSLOG(mwn->parent, LOG_ERR, "NNE writer: mccmnc %d: %s\n", msg->mccmnc,
+                      json_object_to_json_string_ext(obj, JSON_C_TO_STRING_PLAIN));
+
+    if (mwn->metadata_cache == NULL)
+        mwn->metadata_cache = json_object_new_array();
+
+    json_object_array_add(mwn->metadata_cache, obj);
+}
+
+
+/*static void md_nne_process_mode(struct md_writer_nne *mwn,
+                                       struct nne_metadata_descr* descr,
+                                       struct nne_modem* modem,
+                                       struct md_iface_event *mie,
+                                       enum nne_message_source source)
+{
+
+}*/
+
+
+static void md_nne_process_iface_event(struct md_writer_nne *mwn,
+                                       struct nne_metadata_descr* descr,
+                                       struct nne_modem* modem,
+                                       struct md_iface_event *mie,
+                                       enum nne_message_source source)
+{
+    int i;
+    struct nne_value value =  descr->parse_cb(modem, mie);
+
+    if (nne_value_compare(modem->metadata[descr->idx].value, value) != 0) {
```

Comment:
It would be nicer here to check for 0, do the clean up and then return. In that way we will reduce indentation and make code more readable.